File: JerryLi/14_Longest Common Prefix.java
/*
Write a function to find the longest common prefix string amongst an array of strings.
*/

public class Solution {
    public String longestCommonPrefix(String[] strs) {
        /*
        Solution in discussion:
        First sort string array by letter from head to tail.
        Then compare first with last element in sorted string array.
        */
        if (strs==null || strs.length==0) return "";
        Arrays.sort(strs);
        StringBuilder result = new StringBuilder();
        
        char[] first = strs[0].toCharArray();
        char[] last = strs[strs.length-1].toCharArray();
        
        for (int i=0; i<first.length; i++){
            if (last.length > i && last[i]==first[i]){
                result.append(first[i]);
            } else {
                break;
            }
        }
        return result.toString();
    }
}

File: JackChen/string/14. Longest Common Prefix.py
"""
Write a function to find the longest common prefix string amongst an array of strings.
"""

class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        def lcp(str1, str2):
            i, l = 0, min(len(str1), len(str2))
            while i < l:
                if str1[i] == str2[i]:
                    i = i+1
                else:
                    break
            return str1[:i]
        return reduce(lcp, strs) if strs else ''
